### PR TITLE
perf: segment Source content into chunked arrays

### PR DESCRIPTION
**Pull request by rjharmon**
- retains reasonable performance when used for tokenizing large content > 10k bytes

### CODE REVIEW

File: src/tokens/Source.js [hunk 1]
```diff
@@ -11,9 +11,9 @@
 export class Source {
     /**
      * @readonly
-     * @type {string}
+     * @type {string[][]}
      */
-    content
+    contentIndex
```
Comment (christianschmitz) [inline src/tokens/Source.js]: `contentIndex` is a slighly confusing name. Could we rename this to e.g. `segmentedContent` ?

File: src/tokens/Source.js [hunk 2]
```diff
@@ -26,26 +26,57 @@ export class Source {
...
         this.name = options.name ?? "unknown"
     }
 
     /**
      * Number of characters in source content
      * @type {number}
      */
-    get length() {
-        return this.content.length
+    length
```
Comment (christianschmitz) [inline src/tokens/Source.js]: I think all properties should above the constructor, and maybe it makes sense to add the `@readonly` annotation to this property

Sadly I don't know of any formatter that moves properties into the class header automatically

File: src/tokens/Source.js [hunk 3]
```diff
@@ -26,26 +26,57 @@ export class Source {
...
+
+    /**
+     * splits long inputs to segments for more efficient access
+     * @param {string[]} content
+     * @returns {string[][]}
+     */
+    segmentContent(content) {
```
Comment (christianschmitz) [inline src/tokens/Source.js]: I think this utility function shouldn't be a member of this class, but rather be moved into the codec-utils repo, with the segmentSize being a second argument

Also I think a for-loop conveys the intent better:
```ts
const n = content.length
/**
 * @type {string[][]}
 */
const segments = []
    for (let i = 0; i < content.length; i += segmentSize) {
        segments.push(content.slice(i, i + segmentSize));
    }
    return segments;
```

Update: I've added `segmentArray()` to `@helios-lang/codec-utils` v0.1.35, so we can import that here

File: src/tokens/Source.js [hunk 4]
```diff
@@ -26,26 +26,57 @@ export class Source {
      * @param {SourceOptions} options
      */
     constructor(content, options = {}) {
-        this.content = content
+        this.rawContent = content
+        // one-step split to utf-8 runes in the content
+        const codePoints = [...content]
+        // heuristic for segment size
+        this.segmentSize = Math.max(
```
Comment (christianschmitz) [inline src/tokens/Source.js]: For consistency I think all new properties here should get their own block of declarations in the class header above the constructor

File: src/tokens/Source.js [hunk 5]
```diff
@@ -26,26 +26,57 @@ export class Source {
      * @param {SourceOptions} options
      */
     constructor(content, options = {}) {
-        this.content = content
+        this.rawContent = content
```
Comment (christianschmitz) [inline src/tokens/Source.js]: Could you change the name of this property back to `content`? I think I have some downstream packages depend on that

File: src/tokens/Source.js [hunk 6]
```diff
@@ -76,34 +107,45 @@ export class Source {
...
+            c = this.getChar(i)
         }
 
         return chars.join("")
     }
 
+    /**
+     * @type{[number, number][] | undefined}
+     */
+    lineLengths
```
Comment (christianschmitz) [inline src/tokens/Source.js]: I would change the name of this property to `_lineLengths` and make it `@private`, then change its type to `Option<[number, number][]>` and initialize it explicitly in the constructor using `None` imported from @helios-lang/type-utils.

The `calcLineLengths()` function I would change to `get lineLengths()`, which returns either the cached result from the private property, or does the calculation (assigning to the private property in the process).

That way any function that calls `this.lineLengths` doesn't have to worry about the caching.

File: src/tokens/Source.js [hunk 7]
```diff
@@ -76,34 +107,45 @@ export class Source {
...
+    /*
+     * Calculates the length of each line and the total length up to that line
+     * @returns {[number, number][]}
+     */
+    calcLineLengths() {
```
Comment (christianschmitz) [inline src/tokens/Source.js]: See comment about `lineLenghts`

File: src/tokens/Source.js [hunk 8]
```diff
@@ -76,34 +107,45 @@ export class Source {
...
-        let line = 0
-
-        for (let j = 0; j < i; j++) {
-            if (this.content[j] == "\n") {
-                col = 0
-                line += 1
-            } else {
-                col += 1
-            }
+        const lengths = this.lineLengths || this.calcLineLengths()
```
Comment (christianschmitz) [inline src/tokens/Source.js]: See comment above about `lineLengths`

**Review by christianschmitz — CHANGES_REQUESTED** [review]
Thanks for contributing this, I've requested a few changes in order to maintain consistency with the rest of the code-base

File: src/tokens/Source.js (hunk 2 above)
Comment (rjharmon) [reply src/tokens/Source.js]: yup, done

File: src/tokens/Source.js (hunk 1 above)
Comment (rjharmon) [reply src/tokens/Source.js]: made it `contentChunks`

File: src/tokens/Source.js (hunk 3 above)
Comment (rjharmon) [reply src/tokens/Source.js]: ok, thanks.  Using that now.

File: src/tokens/Source.js (hunk 4 above)
Comment (rjharmon) [reply src/tokens/Source.js]: ah, yes - I had thought of that, but then I forgot to move things around before pushing.

thanks!

File: src/tokens/Source.js (hunk 5 above)
Comment (rjharmon) [reply src/tokens/Source.js]: done

File: src/tokens/Source.js (hunk 6 above)
Comment (rjharmon) [reply src/tokens/Source.js]: They're actually line-ending locations; renamed.

The first number in the tuple wasn't actually used/needed, so I changed the type to `number[]` too.

File: src/tokens/Source.js (hunk 7 above)
Comment (rjharmon) [reply src/tokens/Source.js]: thanks for this

File: src/tokens/Source.js (hunk 8 above)
Comment (rjharmon) [reply src/tokens/Source.js]: done

**Comment by rjharmon** [comment]
ok I believe all should be in good shape now.

**Comment by rjharmon** [comment]
ok I believe all should be in good shape now.

**Review by christianschmitz — APPROVED** [review]

**Comment by christianschmitz** [comment]
Thanks for making the requested changes

**Pull request merged**